src/test/java/htsjdk/samtools/cram/encoding/external/ExternalCodecEquivalenceTest.java

```diff
@@ -0,0 +1,176 @@
+import htsjdk.samtools.cram.io.IOTestCases;
+import htsjdk.samtools.util.RuntimeIOException;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Motivation for this test: we were incorrectly encoding a few CRAM record
+ * fields using the wrong data type.  This was surprising, since we would
+ * expect catastrophic "frame shift" type failures from this.
+ *
+ * Instead, we noticed that this conflict causes no problems under the following conditions:
+ *
+ * 1. External Byte, External Integer, and External Long Encodings
+ * create identical output streams when values are limited to the
+ * range 0 to 0x7F because the respective underlying encodings
+ * (raw bytes, ITF8, and LTF8) result in byte-for-byte identical outputs.
+ *
+ * 2. External Integer and External Long Encodings also create identical
+ * output streams when values are limited to 28 unsigned bits
+ * (0 to 0x0F FF FF FF) because ITF8 and LTF8 are equivalent over that range.
+ *
+ * Demonstrate that these conditions do in fact create identical streams.
+ *
+ */
+public class ExternalCodecEquivalenceTest extends HtsjdkTest {
+    // show that External codecs are equivalent within certain ranges
+
+    // if all values are 0 <= v <= 0x7F (high bit 7)
+    // ExternalByteCodec, ExternalIntegerCodec, and ExternalLongCodec streams are identical
+    // because raw bytes, ITF8 and LTF8 are identical
+
+    @Test(dataProvider = "testPositiveByteLists", dataProviderClass = IOTestCases.class)
+    public void byteEquivalenceTest(final List<Byte> values) {
+        codecPairTest(values, this::externalByteCodecWrite,      this::externalIntegerCodecRead);
+        codecPairTest(values, this::externalByteCodecWrite,      this::externalLongCodecRead);
+        codecPairTest(values, this::externalIntegerCodecWrite,  this::externalByteCodecRead);
+        codecPairTest(values, this::externalIntegerCodecWrite,  this::externalLongCodecRead);
+        codecPairTest(values, this::externalLongCodecWrite,     this::externalByteCodecRead);
+        codecPairTest(values, this::externalLongCodecWrite,     this::externalIntegerCodecRead);
+    }
+
+    // if all values are 0 <= v <= 0x0F FF FF FF (high bit 28)
+    // ExternalIntegerCodec, and ExternalLongCodec streams are identical
+    // because ITF8 and LTF8 are identical
+
+    @Test(dataProvider = "testUint28Lists", dataProviderClass = IOTestCases.class)
+    public void intEquivalenceTest(final List<Integer> values) {
+        codecPairTest(values, this::externalIntegerCodecWrite,  this::externalLongCodecRead);
+        codecPairTest(values, this::externalLongCodecWrite,     this::externalIntegerCodecRead);
+    }
+
+    private <T extends Number> void codecPairTest(final List<T> values, Writer<T> writer, Reader reader) {
```

Comment:
It's  a bit confusing to define custom `Reader/Writer` interfaces here since there a really common `java.io` interfaces already with those names.  It's not a big deal either way though.

Reply:
true, will fix